### PR TITLE
Private/mmeeks/bgsave

### PR DESCRIPTION
**Pull request by mmeeks**
A first cut; which should do nothing until you enable it =)

### CODE REVIEW

File: kit/Kit.cpp [hunk 1]
```diff
@@ -1285,6 +1309,170 @@ void Document::updateActivityHeader() const
...
+
+        // hand parentSocket to the main poll
+        KitSocketPoll::getMainPoll()->insertNewSocket(parentSocket);
+        parentWs.reset();
+
+        getLOKit()->setForkedChild(true);
+
+        const auto now = std::chrono::steady_clock::now();
+        LOG_TRC("Background save process " << getpid() << " fork took " <<
+                std::chrono::duration_cast<std::chrono::milliseconds>(now - start).count() << "ms");
```
Comment (Ashod) [inline kit/Kit.cpp:1446]: Chrono serializes with the units already. Removing `.count() << "ms"` should do the right thing and log "ms" as expected).

File: kit/ChildSession.cpp [hunk 2]
```diff
@@ -822,6 +830,38 @@ bool ChildSession::loadDocument(const StringVector& tokens)
     return true;
 }
 
+// attempt to shutdown threads, fork and execute in the background
+bool ChildSession::saveDocumentAsync(const StringVector &tokens)
+{
+#if MOBILEAPP
```
Comment (Ashod) [inline kit/ChildSession.cpp:836]: Perhaps we can do this above, at `static bool doBgSave = !!getenv("COOL_BGSAVE");`. It should make this already-complex function simpler.

File: kit/ChildSession.cpp [hunk 3]
```diff
@@ -543,9 +543,17 @@ bool ChildSession::_handleInput(const char *buffer, int length)
             }
             else if (tokens[1].find(".uno:Save") != std::string::npos)
             {
-                // Disable processing of other messages while saving document
-                InputProcessingManager processInput(getProtocol(), false);
-                return unoCommand(tokens);
+                static bool doBgSave = !!getenv("COOL_BGSAVE");
```
Comment (Ashod) [inline kit/ChildSession.cpp:546]: Would we have a separate command for sync-saving when cleaning up DocBroker?

File: kit/ChildSession.cpp [hunk 4]
```diff
@@ -543,9 +543,17 @@ bool ChildSession::_handleInput(const char *buffer, int length)
...
+
+                // Attempt to save in the background.
+                if (doBgSave && !saveDocumentAsync(tokens))
+                { // fallback to synchronous save
+
+                    // Disable processing of other messages while saving document
```
Comment (Ashod) [inline kit/ChildSession.cpp:552]: I thought ".uno:Save" returned after queuing it, and it was handled in the background. Perhaps I misunderstand how saving is handled in Core.

File: kit/ChildSession.hpp [hunk 5]
```diff
@@ -90,6 +90,13 @@ class ChildSession final : public Session
         return _docManager->sendFrame(msg.data(), msg.size(), WSOpCode::Text);
     }
 
+    bool sendSaveFrame(const std::string& msg)
+    {
+        bool ret = sendTextFrame(msg);
+        _docManager->handleSaveMessage(msg);
```
Comment (Ashod) [inline kit/ChildSession.hpp:96]: Isn't it simpler to just do this line above, at line 3007, if `saveCommand` is true, after we `sendTextFrame`? Unless we expect this to be called from multiple places, or to become more complex, that is.

**Review by Ashod — APPROVED** [review]
Thanks, @mmeeks. Good stuff. I guess we'll enable this under 'experimental' at some point. Would be interesting to see it gets exercised in our tests. Left some questions/suggestions.

File: kit/Kit.cpp (hunk 1 above)
Comment (mmeeks) [reply kit/Kit.cpp:1446]: oh nice =) fixed.

File: kit/ChildSession.cpp (hunk 3 above)
Comment (mmeeks) [reply kit/ChildSession.cpp:546]: Ah perhaps - we want a parameter here too for the save uno command to enable this code-path I think; TBD.

File: kit/ChildSession.cpp (hunk 4 above)
Comment (mmeeks) [reply kit/ChildSession.cpp:552]: Hmm; synchronous save is perhaps misleading though it mostly happens in that callback - the final status thing comes later (I think).

File: kit/ChildSession.hpp (hunk 5 above)
Comment (mmeeks) [reply kit/ChildSession.hpp:96]: Sure - why not; fixed.

**Comment by Ashod** [comment]
Thanks for the update, @mmeeks. Good improvements.

**Pull request merged**